Document distributions in the 6.1 upgrade guide. (#1733)
plone/documentation: docs/backend/upgrading/version-specific-migration/upgrade-to-61.md

Reviewer: I'll edit this file later today. Please hold on merging until I get the chance.

-   If the `plone.app.discussion` Python package is _not_ in your setup, but the site has existing comments (discussions), then the migration code stops with an error.
    Apparently you _were_ using comments in your site.
    Add the `plone.app.discussion` package to your dependencies, and run the migration again.

## Distributions

Plone 6.1 introduces the concept of a Plone {term}`distribution`.
A Plone distribution is a Python package that defines specific features, themes, add-ons, and configurations that get activated when creating a Plone site.

> Now it is available in core Plone as the recommended way for {doc}`creating a new Plone site </install/add-site>`.

Reviewer: I found no way to actually run `plone.distribution` at the suggested page, as the target states:

```
assumes that you have already {doc}`installed </install/index>` and started Plone.
```

...and none of those options pull in a 6.1 site, and neither in the currently open PRs for docs.

What did I miss?

Reviewer 2: @stevepiercy You don't "run plone.distribution" per se. It's just there if you install the `Plone` package (which is what all the documented install methods do) rather than the more minimal `Products.CMFPlone` package.

It's by design that the install methods as documented still use Plone 6.0, since Plone 6.1 is still pre-final release.

With the Cookieplone project template, be sure to choose Plone 6.1.0b1:
```
  [7/17] Should we use prerelease versions? (No): Yes
  [8/17] Plone Version (6.1.0b1): 
```

With buildout- or pip-based installs, be sure to use `https://dist.plone.org/release/6.1-latest/` instead of `https://dist.plone.org/release/6-latest/`

Reviewer: I've updated the checklist for updates to docs for Plone 6.1 final:

https://github.com/plone/documentation/issues/1721

```{seealso}
For more information, see {doc}`/conceptual-guides/distributions`.
```

Distributions are optional.
If your project only uses the `Products.CMFPlone` Python package, you can still create a Plone site in the old way.

> Consider, however, the following differences from distributions and from Plone 6.0.

Reviewer: This sentence confused the heck out of me, and I only understood its intent after reading the list and already knowing some differences. I rewrote it for clarity. I hope it makes more sense now.

-   The configuration form is simpler and shorter.
-   The created site has no content, and therefore no {guilabel}`News` or {guilabel}`Events` folders.
-   You must activate add-ons through the {guilabel}`Add-ons` control panel.

There are a few things you should consider when upgrading a project to, or making an add-on compatible with, Plone 6.1:

-   In general, you don't need to change anything.
    Your existing site will keep working.
    But adding a new site may change in the ways described earlier.
-   Do you want to use the `Products.CMFPlone` package (0 distributions), either `plone.volto` or `plone.classicui` (1 distribution), or `Plone` (2 distributions)?
-   If your site uses the default Volto frontend, you will already have `plone.volto` as a dependency.
    This can stay the same.

> -   If your site uses only the `Products.CMFPlone` package, the frontend is Classic UI.

Reviewer: I thought all Plone sites use `Products.CMFPlone`? Should this be:

```suggestion
-   If your site uses Classic UI for the frontend, you will already have `Products.CMFPlone` as a dependency and not `plone.volto`.
```

Reviewer 2: This line is talking about the situation where your project _only_ installs `Products.CMFPlone` and not `Plone`. Your suggestion as worded is not quite the point Maurits was trying to make here. Instead I suggest:
```suggestion
-   If your site depends on the `Products.CMFPlone` package without the `Plone` or `plone.volto` packages, then the frontend is Classic UI.
```

A Plone distribution is a Python package that defines specific features, themes, add-ons, and configurations that get activated when creating a Plone site.
Now it is available in core Plone as the recommended way for creating a Plone site.

When you install the [`Plone` 6.1 Python package](https://pypi.org/project/Plone/), you get several new packages and two distributions:

-   [`plone.distribution`](https://github.com/plone/plone.distribution) is the new main package that offers the basis for distributions.
-   [`plone.exportimport`](https://github.com/plone/plone.exportimport) imports and exports content, users, and other objects between Plone sites and when upgrading Plone.
    It is now the preferred method, and `plone.distribution` uses it.
-   [`plone.volto`](https://github.com/plone/plone.volto) is the distribution to create a Plone site with the default frontend Volto.
    This package was already present in Plone 6.0, but was upgraded to a distribution.
-   [`plone.classicui`](https://github.com/plone/plone.classicui) is a new package and is the distribution to create a Plone site with the Classic UI frontend.

> When you start a new project with `Plone` 6.1, the launch screen prompts you to choose a frontend distribution when creating a new Plone site.

Reviewer: We need to document the steps after `pip install Plone==6.1.0a5` (latest alpha version to try) and before this screenshot.

I have no idea how to get to this screenshot from this upgrade guide. It should go into the [Install documentation](https://6.docs.plone.org/install/), and add a link to that from here.

Also where does this fit into Cookieplone?

Author: Start Plone (`bin/instance fg`, `docker compose up`, `make start` or whatever) and go to localhost:8080.
Calling it the launch screen helps.

I would say: if someone is reading the upgrade guide, then they already have a Plone site and know how to start it, so we don't need to describe it here.

Well, "start a new project" is probably a bad choice of words when what I really mean is just "start Plone".  Let me change that.

Author: O, I see here and in a few other lines that you call it "frontend distributions".  That is not what they are, they are distributions installed on the backend.  It is just that in the default case your choice of distribution will be guided by what frontend you want.  If you are going to use the Volto frontend, you must install the `default` distribution (plone.volto).  If you are going to use Classic UI, you must install the `classic` distribution.

The tricky thing of course is that the `classic` distribution gives you a frontend as well, as this is the server side rendered frontend that is baked into the backend...

A Plone agency might very well love Classic UI and have a demo site where at startup you can choose between five distributions that are all variations of the `classic` distribution: website, intranet, extranet, web shop, university, etc.

I have renamed "frontend distribution" to "distribution" everywhere.

Reviewer: Well, the thing is, Classic UI is a frontend, even though it is generated on the backend, just as Node.js generates the frontend for Volto.

> Well, "start a new project" is probably a bad choice of words when what I really mean is just "start Plone". Let me change that.

Yes, that's what threw me. We're on the same page now for this item.

```{image} images/distribution-chooser.png
:alt: screenshot of the launch screen
```

After you select a frontend distribution, you will fill out a brief form to configure your new Plone 6.1 site, similar to the process for Plone 6.0.

Distributions are optional.
If your project only uses the `Products.CMFPlone` Python package, you can still create a Plone site in the usual way.
Consider, however, the following differences from distributions.

-   The configuration form is simpler and shorter.
-   The created site has no content, and therefore no {guilabel}`News` or {guilabel}`Events` folders.

> -   You must install then activate add-ons through the {guilabel}`Add-ons` control panel.

Reviewer: The original paragraph was unclear to me, so I attempted to clarify it with a shortened paragraph and bullet points. Please let me know if this works for y'all.

Author: Yes, that works.
But what I tried to do was apparently not clear.  I did not want to describe the difference between CMFPlone 6.1 and a distribution, but the difference between CMFPlone 6.0 and 6.1. 

* In CMFPlone 6.0 and 6.1 the standard form is the same, except that in 6.1 the option for install the default content is gone.  That is covered by bullet two.
* In 6.0 you had the option of going to an advanced form where you could select a dozen or so extra addons/profiles for installation.  This advanced form is gone in 6.1.  This is basically covered by your last bullet point.  That bullet point does not mention the missing advanced form, but that may be fine.

I have slightly modified the text so I think it now correctly describes both differences: CMFPlone 6.1 compared to distributions and compared to CMFPlone 6.0.

-   If the `plone.app.discussion` Python package is _not_ in your setup, but the site has existing comments (discussions), then the migration code stops with an error.
    Apparently you _were_ using comments in your site.
    Add the `plone.app.discussion` package to your dependencies, and run the migration again.

## Distributions

Plone 6.1 introduces the concept of a Plone {term}`distribution`.
A Plone distribution is a Python package that defines specific features, themes, add-ons, and configurations that get activated when creating a Plone site.
Now it is available in core Plone as the recommended way for creating a Plone site.

> When you install the [`Plone` 6.1 Python package](https://pypi.org/project/Plone/), you get several new packages and two distributions:

Reviewer: We need to reference how someone may install Plone 6.1. I would suggest something like this, since this is all we have at this moment. It's bad because it doesn't cover Classic UI.

```suggestion
When you [install the `Plone` 6.1 Python package](https://6.docs.plone.org/install/), you get several new packages and two distributions:
```

From there, we have [Create a project with Volto (development or pre-release)](https://6.docs.plone.org/install/create-project-cookieplone.html) (a very recent addition) but nothing for Classic UI for Plone 6.1. 😱 

----

As a side note, IMO, this is a critical problem that should block the release of Plone 6.1 final. Installation of Plone 6.1 and major development blockers—specifically how to write tests, create an add-on, and customize a theme—should be resolved. I don't know if anyone shares this opinion.

-   [`plone.distribution`](https://github.com/plone/plone.distribution) is the new main package that offers the basis for distributions.
-   [`plone.exportimport`](https://github.com/plone/plone.exportimport) imports and exports content, users, and other objects between Plone sites.
    `plone.distribution` uses it.
-   [`plone.volto`](https://github.com/plone/plone.volto) is the distribution to create a Plone site with the default frontend Volto.
    This package was already present in Plone 6.0, but was upgraded to a distribution.
-   [`plone.classicui`](https://github.com/plone/plone.classicui) is a new package and is the distribution to create a Plone site with the Classic UI frontend.

When you start Plone, the launch screen prompts you to choose a distribution when creating a new Plone site.

````{card}
```{image} /backend/upgrading/version-specific-migration/images/distribution-chooser.png
:alt: Launch screen for choosing a distribution
:target: /_images/distribution-chooser.png
```
+++
_Launch screen for choosing a distribution_
````

> After you select a distribution, you will fill out a brief form to configure your new Plone 6.1 site, similar to the process for Plone 6.0.

Reviewer: I'm reconsidering lines 137-148 for inclusion in the Upgrade Guide. I think they belong instead in one of the [Install](https://6.docs.plone.org/install/) guides, but I'm not sure which, if any of them. I think the UI is going to change with future 6.1.x, and we almost never update an Upgrade Guide after its release, so I don't want these details and screenshot to become stale.

Distributions are optional.

> If your project only uses the `Products.CMFPlone` Python package, you can still create a Plone site in the usual way.

Reviewer: We need to clarify what is "the usual way". plone-cookie-starter, buildout, or ...?

Consider, however, the following differences from distributions and from Plone 6.0.

-   The configuration form is simpler and shorter.
-   The created site has no content, and therefore no {guilabel}`News` or {guilabel}`Events` folders.
-   You must activate add-ons through the {guilabel}`Add-ons` control panel.

There are a few things you should consider when upgrading a project to, or making an add-on compatible with, Plone 6.1:

> -   In general, you don't need no change anything.

Reviewer: ```suggestion
-   In general, you don't need to change anything.
```

    Your existing site will keep working.
    But adding a new site may change in the ways described earlier.
-   Do you want to use the `Products.CMFPlone` package (0 distributions), either `plone.volto` or `plone.classicui` (1 distribution), or `Plone` (2 distributions)?

> -   When your site used the default Volto frontend, you will already have `plone.volto` as a dependency.

Reviewer: ```suggestion
-   When your site uses the default Volto frontend, you already have `plone.volto` as a dependency.
```

    This can stay the same.
-   When your site uses only the `Products.CMFPlone` package, the frontend is Classic UI.
    This can stay the same, but you may want to depend on `plone.classicui`.
    With that package you can still create a new site and have the same content as before.

> -   When your site is using the `Plone` package, you will have the two new distributions available.

Reviewer: ```suggestion
-   When your site uses the `Plone` package, you have the two new distributions available.
```

    This is fine.

>     If you know you only need `plone.volto` or only need `plone.classicui`, you can switch to that.

Reviewer: ```suggestion
    If you know you only need `plone.volto`, or only need `plone.classicui`, you can switch to only the one.
```

    You can also limit the options for selecting a distribution by setting the environment variable `ALLOWED_DISTRIBUTIONS` with fewer options.
    Set `ALLOWED_DISTRIBUTIONS=default` for the distribution targeting the default Volto frontend (`plone.volto`).
    Set `ALLOWED_DISTRIBUTIONS=classic` for the distribution with the Classic UI frontend (`plone.classicui`).
-   If you switch from `Plone` to `plone.volto` or `plone.classicui`, you might want to install extra core add-ons, for example `plone.app.upgrade` or `plone.app.caching`.

> -   If your add-on is only for Volto, you may want to add `plone.volto` as a dependency.

Reviewer: ```suggestion
-   If your add-on is only for Volto, you might want to add `plone.volto` as a dependency.
```

> -   If your add-on is only for Classic UI, you may want to add `plone.classicui` as a dependency.

Reviewer: ```suggestion
-   If your add-on is only for Classic UI, you might want to add `plone.classicui` as a dependency.
```

    Note though that `plone.classicui` is not available for Plone 6.0.

You can create your own distributions to suit your needs.

-   Create a distribution demonstrating your favorite setup for Plone.
    This would contain the add-ons that you usually add in each project, including example content.
    With this you can create a fully configured Plone site filled with content for a potential client.
    Or they can try it themselves if the distribution is available publicly.
-   Create a distribution for an internal or client project.
    This would create a site with all add-ons activated and configured for this project, including example content, and optionally users and groups.
    During the development phase of a new project, all developers would use this to create a fresh site locally so that everyone has the same configuration and content.
    Before the project goes live, you can use the distribution to create the initial setup.

```{note}
For Plone 7, the [Plone roadmap](https://plone.org/why-plone/roadmap) guides development toward a clearer separation between the Classic UI frontend and the core `Products.CMFPlone` backend.

> This means in Plone 7 `Products.CMFPlone` will have less code and pull in fewer dependencies.

Reviewer: ```suggestion
This means in Plone 7, `Products.CMFPlone` will have less code and pull in fewer dependencies.
```
